Project External Links Block

srallen opened the pull request:
Staging branch URL: https://project-external-links.pfe-preview.zooniverse.org

Related to #4371. We discussed trying to move all external links into its own div block similar to organization home pages. This is a early version of this design idea.

@beckyrother I'll need some design guidance how this looks on organization home pages and project home pages:
Snapshot Safari organization: https://project-external-links.pfe-preview.zooniverse.org/organizations/meredithspalmer/snapshot-safari?env=production
Snapshot Godwana (a project with a lot of external links): https://project-external-links.pfe-preview.zooniverse.org/projects/meredithspalmer/snapshot-gondwana?env=production

There are slight differences in the project and organization designs in terms of headers and font sizes. One way to resolve this would be to reuse this new ExternalLinks component between the project and organization and make sure the font sizes and headers match between the home pages.

Another issue is if the researcher quote and links area are both visible, then the layout doesn't really look good since the quote gets smashed a bit. This can be seen on the Snapshot Godwana. Reworking projects to be like organizations with its content layout isn't going to work 1:1 since project introductions are typically shorter and are in the project hero. Organizations have their introductions further down the page and paired with the quote from the researcher. Let me know how you'd like to resolve this.

One last thing, there was an inconsistency with the organization links. In the project hero, it was treated as an internal link as it should be, since it doesn't go off site. In the navbar it was treated as an external link that opened a new tab and had the external link icon. I changed the org link in the navbar behave like an internal link. 

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [x] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [x] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [x] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

beckyrother commented:
I'd like to make sure Organizations is visually different from the projects within it, so it would be good to keep the differences in the header – is that what you're asking? 

However, I do agree to reuse the ExternalLinks component so it looks the same across everything. This is a long quote on the Gondwana project and I'm not sure how to resolve that, unless the modules were evenly split 1:1:1 rather than proportionally as they are now. Maybe try that and see if it helps?

srallen commented:
@beckyrother this is ready for another design review. 

- What an organization with external links looks like: https://project-external-links.pfe-preview.zooniverse.org/organizations/markb-panoptes/nfnorgtest
- An organization without external links: https://project-external-links.pfe-preview.zooniverse.org/organizations/markb-panoptes/urbanwildlifeorgtest
- A project with only external links: https://project-external-links.pfe-preview.zooniverse.org/projects/rafe-dot-lafrance/notes-from-nature-on-staging
- A project with only a researcher quote: https://project-external-links.pfe-preview.zooniverse.org/projects/srallen086/workflow-assignment-testing
- A project with both: https://project-external-links.pfe-preview.zooniverse.org/projects/srallen086/srallen-testing
- A project with none: https://project-external-links.pfe-preview.zooniverse.org/projects/srallen086/galaxy-zoo-in-astronomy-101

srallen commented:
Note to self: The project about section isn't wrapping properly on smaller screen sizes anymore.

srallen commented:
This is ready for technical review. A few things to note in addition to the above design comments:
- I did a organizational refactor to put the `ExternalLinksBlock` into the top level components folder and put the helpers into the lib folder and the test helpers into the top level test folder. This is because the component is now shared between the organization page and project home page.
- I removed the redundancy between the `social-links.js` and `socialIcons.json` file and preferred the later. The social links editor in the lab had a tiny refactor because of it.
- I fixed some incorrect CSS with the researcher quote css. It was using both flex and block layout floats, which according to spec don't work together. It's been refactored to properly use flex. 
- Acknowledging here that IE 11 does some unexpected wrapping with the ExternalLinksBlock, but I don't think it's worth the effort in fixing since it doesn't really look bad, in my opinion. IE 11 has a lot of known flex bugs and we should encourage folks to use Edge instead of continuing to fix every IE 11 bug we see.

rogerhutchings commented:
I had a few observations on the design first (some of these may be nitpicking):

**Project external links block**

<img width="296" alt="screenshot 2018-03-06 11 27 11" src="https://user-images.githubusercontent.com/2725841/37030125-aaa0d890-2131-11e8-9bea-1f83e7e06565.png">

**Org external links block**

<img width="341" alt="screenshot 2018-03-06 11 25 58" src="https://user-images.githubusercontent.com/2725841/37030120-a7e844f8-2131-11e8-876a-5a89f2153b6a.png">

**Observations**

- Heading font size is very small on the org version.
- Font family for the block is Open Sans on the project, Karla on the org (although I appreciate that the project page still needs updating).
- This might be a personal thing, but the underline on hover under the icon looks a bit ugly, and isn't always visible depending on the icon and browser. The screenshots were taken on Chrome, and the underline isn't visible at all on Chrome or Safari. It _is_ visible on Firefox, but it's right up against the underside of the icon.
- Icons are right-aligned, so wider icons like G+ look a bit offset.

rogerhutchings commented on app/components/ExternalLinksBlock/ExternalLinksBlock.jsx:
```diff
@@ -0,0 +1,81 @@
+    light: zooTheme.colors.navy.default
+  })};
+  flex: ${props => `1 0 ${props.basis}%`};
+  min-width: 400px;
```
What about using the `pxToRem()` function? [(example)](https://github.com/zooniverse/Panoptes-Front-End/blob/e07c34edad9d1ea5773781453ce4d3d689567ddc/app/pages/project/components/ProjectNavbar/components/ProjectNavbarNarrow/NarrowMenuButton.jsx)

rogerhutchings replied:
We should probably lay down some rules on when to use which unit as well...

srallen replied:
This is to enforce a specific layout, so pixels make sense to me here. Having some rules on fixed vs responsive unit usage would be useful.

rogerhutchings commented on app/components/ExternalLinksBlock/ExternalLinksBlock.jsx:
```diff
@@ -0,0 +1,81 @@
+  return (
+    <ThemeProvider theme={{ mode: 'light' }}>
+      <StyledExternalLinksBlock basis={basis}>
+        {children}
```
This is only ever going to be a title, so it might be better to pass the title content as a prop and markup/style it here.

eatyourgreens commented on app/components/ExternalLinksBlock/ExternalLinksBlockContainer.jsx:
```diff
@@ -0,0 +1,47 @@
+import ExternalLinksBlock from './ExternalLinksBlock';
+
+export default class ExternalLinksBlockContainer extends React.Component {
+  getExternalLinks() {
```
They've never needed to be bound, except for event handlers. I've been asking for them to removed. They don't do anything, as far as I can tell, except create a new, identical copy of the original method. It can be hard to debug whether test spies and stubs are wrapping the original or the copy.

Re. event handlers, I noticed that the Facebook codemod, which we've used to generate a few ES6 components, prefers to bind in the JSX render: `onClick={this.handleClick.bind(this_}`

rogerhutchings replied:
Huh, today I learned. Thanks @eatyourgreens, I must have misunderstood what was required when React moved away from `createClass`

eatyourgreens replied:
I've only been picky about this since we upgraded to enzyme 3. Tests that spy on `instance.method` have problems now if `method` has been bound to a new context.

They've always puzzled me since class methods are bound to instances of the class by default. Event handlers and promise callbacks invoke functions with a new `this` context, if you don't bind them. Beyond those two cases, I think it should be okay not to use them.

rogerhutchings replied:
[I think this is what caught me out](https://reactjs.org/blog/2015/01/27/react-v0.13.0-beta-1.html#autobinding):

> React.createClass has a built-in magic feature that bound all methods to this automatically for you. This can be a little confusing for JavaScript developers that are not used to this feature in other classes, or it can be confusing when they move from React to other classes.

> **Therefore we decided not to have this built-in into React’s class model.** You can still explicitly prebind methods in your constructor if you want.

(emphasis mine)

There wasn't a mention on when you would and wouldn't need to do it, so I probably assumed it needed to be done ¯\_(ツ)_/¯

srallen commented:
@rogerhutchings to answer the design questions:
- @beckyrother asked for visual differentiation in the headers between orgs and projects and confirmed in this comment: https://github.com/zooniverse/Panoptes-Front-End/pull/4402#issuecomment-368962189
- Font family should be updated, but I'd probably do that in a separate PR

For the last two points, we can check in with what @beckyrother would like. I added the underline because the org version was missing hover, focus states entirely, so it was a quick and easy way to make them more accessible. I don't personally have a preference as to what the hover, focus state is. Perhaps border-bottom instead so it's more visually prominent?

srallen commented on app/components/ExternalLinksBlock/ExternalLinksBlock.jsx:
```diff
@@ -0,0 +1,81 @@
+    padding-left: 0;
+  }
+
+  > ul li {
```
Yeah it is. I can nest the styles to make it a bit more readable.

beckyrother commented:
- Extremely agree about the font – should definitely be switched to Karla and make sure the size is the same. 

- We (and most of the internet) treat every other text link with an underline, so for consistency's sake please keep it an underline. If you'd like to make it border-bottom so it's thicker and matches the nav bar, that's fine, but it needs to be an underline.

- Not a lot we can do about the G+ icon, that's the width that it is. I prefer the icons to be right-aligned so there's a consistent column of white space between icons and their labels.

rogerhutchings commented:
Thanks @srallen & @beckyrother! I'm not questioning underlining links, just the icon as it stands. For what it's worth, I quickly mocked up a couple of alternatives in-browser:

`border-bottom`:

<img width="228" alt="screenshot 2018-03-06 21 22 12" src="https://user-images.githubusercontent.com/2725841/37059404-44d60d74-2185-11e8-835a-959399120d71.png">

Underline on the text only:

<img width="193" alt="screenshot 2018-03-06 21 26 34" src="https://user-images.githubusercontent.com/2725841/37059421-539a5752-2185-11e8-824a-fc68a4414046.png">

I'd prefer either to the current one, but this is a purely subjective thing and ultimately not my decision 😁🤐

beckyrother commented:
@rogerhutchings Oh! I completely misunderstood, my mistake. You're totally right, the underline on the icon looks silly. The second option you provide is great! Happy to move forward with that one.

srallen commented:
@rogerhutchings this is ready for another review

srallen commented:
@rogerhutchings do you have a moment to do a final review on this? It's been rebased with the latest master. Thanks!

rogerhutchings commented:
LGTM 👍

beckyrother commented:
Should this have a grey background? Looking at Planet Four: Terrains, it's really awkward below the "1 person is talking about this" https://project-external-links.pfe-preview.zooniverse.org/projects/mschwamb/planet-four-terrains

![image](https://user-images.githubusercontent.com/1392130/37350230-3d0130ee-26a6-11e8-9b6d-2c9a5d224173.png)

srallen commented:
@beckyrother do you mean the external links block area to have the grey background?

beckyrother commented:
@srallen I'm sorry to keep adding things. Is that screenshot representative of the layout project pages will usually have? if it is, yes, it should have the same grey background as behind the stats. If not white is fine.

srallen commented:
@beckyrother The launched approved projects will likely have commenters on Talk, so I'll make the background grey. Do you want it grey or white for organizations?

beckyrother commented:
@srallen Sorry, for organizations it should stay white.

srallen commented:
@beckyrother I've updated the background colors, however, there seems to be even more variability here. The talk stat sometimes is next to talk subject images above the project stats instead of next to the project stats. Planet Four: Terrains loaded with the talk stats next to the project stats, because here how it looks for me with the grey background:

<img width="1437" alt="screen shot 2018-03-15 at 9 38 28 am" src="https://user-images.githubusercontent.com/5016731/37470066-abf757be-2834-11e8-9e6c-9edd3e1935a8.png">

I think the variability depends on the type of Talk comment, if it's for a subject or not maybe? Do you want to continue with this change? It looks ok in the other cases:

Staging organization with white background: http://localhost:3735/organizations/markb-panoptes/nfnorgtest
Staging project with grey background and the talk stat next to the project stats:
http://localhost:3735/projects/srallen086/srallen-testing

beckyrother commented:
Nothing is ever simple! Let's leave it grey like this, I like the white-teal-grey in the image you posted. Approved on my end! Thank you Sarah :)

srallen commented:
@beckyrother Ok! Should this merge or would you like to send the staged branch link out to the test listserv?

beckyrother commented:
@srallen we should send this to the listserve again. I added a few more people to the spreadsheet and I'm so sorry but I'm not able to get into the mailing list to edit it. I will write a little thing and link to both my spreadsheet and the email so they can go out all at once though.

srallen commented:
@beckyrother ok sounds good. Let's keep the spreadsheet link off of github just for privacy since it contains email addresses.

beckyrother commented:
@srallen here's the email: https://docs.google.com/document/d/1O89hlcOktKhMhcWM-lR4vBr_lwejU-fPE6ud5WMzpMs/edit?usp=sharing
and i'll dm you the updated emails

simoneduca commented:
Is this waiting on some more feedback from the email that was sent?

beckyrother commented:
Yes, I'm planning to leave the survey open till Friday, so we can merge Monday morning. 

Feedback:
Would it be possible to allow a project owner to specify that a single link stay in the nav bar? Meg says her field guide information is in an external link, and I'm not sure how many other projects do that but I don't want that information to be lost. Interested in hearing thoughts about this idea.

srallen commented:
@beckyrother that's an edge case and an unlikely one for most projects. Planet Four: Terrains launched before we implemented field guides in the project builder, so this project is linking back to a page on the original Planet Four. The original Planet Four is also in the process of being migrated to Panoptes, so regardless, she'll need to migrate the Planet Four: Terrains field guide to the project builder version once the old Planet Four shuts down. I don't think we should come up with a solution just for this scenario of an external link and advise migrating it to the project builder instead.

beckyrother commented:
@srallen I completely agree about her information, she needs to migrate it to our Field Guide. 

We've only gotten 4 responses so far but there haven't been any other actionable requests. I'll post here again if that changes.

beckyrother commented:
@srallen got one good comment from the survey: This section should be called "External Project Links" rather than just "Links" 
That's all! Survey closes at 5 today but I think we're ready to go after that minor text update.
